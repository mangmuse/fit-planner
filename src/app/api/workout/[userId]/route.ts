import { getWorkouts } from "@/app/api/_utils/getWorkouts";
import { handleServerError } from "@/app/api/_utils/handleError";
import { validateData } from "@/util/validateData";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";

export const GET = async (
  req: NextRequest,
  { params }: { params: Promise<{ userId: string }> }
) => {
  try {
    const { userId } = await Promise.resolve(params);
    const parsedUserId = validateData<string>(z.string(), userId);
    const workouts = await getWorkouts(parsedUserId);

    return NextResponse.json({ success: true, workouts }, { status: 200 });
  } catch (e) {
    return handleServerError(e);
  }
};
