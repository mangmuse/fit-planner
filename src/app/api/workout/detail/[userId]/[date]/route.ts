import { prisma } from "@/lib/prisma";
import { NextRequest, NextResponse } from "next/server";

export async function GET(
  req: NextRequest,
  { params }: { params: Promise<{ userId: string; date: string }> }
) {
  const { userId, date } = await Promise.resolve(params);
  const parsedDate = new Date(date);
  const workout = await prisma.workout.findFirst({
    where: { userId, date: parsedDate },
  });

  if (!workout) {
    return NextResponse.json([], { status: 200 });
  }

  const workoutDetails = await prisma.workoutDetail.findMany({
    where: { workoutId: workout.id },
    orderBy: [{ exerciseOrder: "asc" }, { setOrder: "asc" }],
    include: {
      exercise: {
        select: {
          name: true,
        },
      },
    },
  });

  const flattenedWorkoutDetails = workoutDetails.map((detail) => ({
    ...detail,
    exerciseName: detail.exercise.name,
    exercise: undefined,
  }));
  console.log(flattenedWorkoutDetails);
  return NextResponse.json(flattenedWorkoutDetails, { status: 200 });
}
