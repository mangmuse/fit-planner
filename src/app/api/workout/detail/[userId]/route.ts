import { getWorkoutIds, getWorkouts } from "@/app/api/_utils/getWorkouts";
import { handleServerError } from "@/app/api/_utils/handleError";
import { prisma } from "@/lib/prisma";
import { validateData } from "@/util/validateData";
import { WorkoutDetail } from "@prisma/client";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";

export const GET = async (
  req: NextRequest,
  { params }: { params: Promise<{ userId: string }> }
) => {
  try {
    const { userId } = await Promise.resolve(params);
    const parsedUserId = validateData<string>(z.string(), userId);
    const workoutIds = await getWorkoutIds(parsedUserId);
    const workoutDetails = await prisma.workoutDetail.findMany({
      where: {
        workoutId: {
          in: workoutIds,
        },
      },
      include: {
        exercise: {
          select: {
            name: true,
          },
        },
      },
    });
    const transformedDetails: WorkoutDetail[] = workoutDetails.map((detail) => {
      const { exercise, ...rest } = detail;
      return { ...rest, exerciseName: exercise.name };
    });
    return NextResponse.json({
      success: true,
      workoutDetails: transformedDetails,
    });
  } catch (e) {
    return handleServerError(e);
  }
};
