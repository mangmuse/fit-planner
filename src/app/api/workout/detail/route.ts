import { convertDatesToStrings } from "@/app/api/_utils/dateConverter";
import { getWorkoutIds } from "@/app/api/_utils/getWorkouts";
import { handleServerError } from "@/app/api/_utils/handleError";
import { prisma } from "@/lib/prisma";
import { ClientWorkoutDetail } from "@/types/models";
import { validateData } from "@/util/validateData";
import { WorkoutDetail, Prisma } from "@prisma/client";
import { NextRequest, NextResponse } from "next/server";
import { z } from "zod";

export type WorkoutDetailWithIncludes = Prisma.WorkoutDetailGetPayload<{
  include: {
    exercise: {
      select: {
        name: true;
      };
    };
  };
}>;
export const GET = async (req: NextRequest) => {
  try {
    const userId = req.nextUrl.searchParams.get("userId");
    if (!userId || typeof userId !== "string") {
      return NextResponse.json(
        { success: false, message: "userId가 없거나 타입이 올바르지 않습니다" },
        { status: 400 }
      );
    }
    const parsedUserId = validateData<string>(z.string(), userId);
    const workoutIds = await getWorkoutIds(parsedUserId);
    if (workoutIds.length === 0) {
      return NextResponse.json({ success: true, workoutDetails: [] });
    }
    const workoutDetails = await prisma.workoutDetail.findMany({
      where: {
        workoutId: {
          in: workoutIds,
        },
      },
      include: {
        exercise: {
          select: {
            name: true,
          },
        },
      },
    });
    const transformedDetails: ClientWorkoutDetail[] = workoutDetails.map(
      (detail: WorkoutDetailWithIncludes) => {
        const { exercise, ...rest } = detail;
        return {
          ...rest,
          exerciseName: exercise.name,
          createdAt: rest.createdAt.toISOString(),
          updatedAt: rest.updatedAt?.toISOString(),
        } as ClientWorkoutDetail;
      }
    );
    return NextResponse.json({
      success: true,
      workoutDetails: transformedDetails,
    });
  } catch (e) {
    return handleServerError(e);
  }
};

export async function POST(req: NextRequest) {
  const { userId, selectedExercises, date } = await req.json();

  let workout = await prisma.workout.findFirst({
    where: {
      userId,
      date: new Date(date),
    },
  });

  let startOrder = 1;

  if (!workout) {
    workout = await prisma.workout.create({
      data: { userId, date: new Date(date) },
    });
  } else {
    const lastDetail = await prisma.workoutDetail.findFirst({
      where: { workoutId: workout.id },
      orderBy: { exerciseOrder: "desc" },
    });
    startOrder = lastDetail ? lastDetail.exerciseOrder + 1 : 1;
  }

  const details = selectedExercises.map(
    (exerciseId: number, index: number) => ({
      workoutId: workout.id,
      exerciseId,
      exerciseOrder: startOrder + index,
      setOrder: 1,
    })
  );

  await prisma.workoutDetail.createMany({
    data: details,
  });

  return Response.json({ success: true });
}
