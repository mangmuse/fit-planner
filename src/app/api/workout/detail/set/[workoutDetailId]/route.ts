import { NextRequest, NextResponse } from "next/server";
import { prisma } from "@/lib/prisma";

export async function DELETE(
  request: NextRequest,
  { params }: { params: Promise<{ workoutDetailId: string }> }
) {
  const { workoutDetailId } = await Promise.resolve(params);

  try {
    await prisma.workoutDetail.delete({
      where: { id: workoutDetailId },
    });
    return new NextResponse(null, { status: 204 });
  } catch (error) {
    console.error(error);
    return new NextResponse("세트 삭제에 실패했습니다", { status: 500 });
  }
}
