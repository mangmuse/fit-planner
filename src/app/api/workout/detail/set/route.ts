import { prisma } from "@/lib/prisma";
import { NextRequest, NextResponse } from "next/server";

export async function POST(req: NextRequest) {
  const newSet = await req.json();
  if (!newSet) {
    return NextResponse.json(
      { error: "세트 추가에 실패했습니다" },
      { status: 400 }
    );
  }
  await prisma.workoutDetail.create({ data: newSet });
  return NextResponse.json({ success: true });
}
