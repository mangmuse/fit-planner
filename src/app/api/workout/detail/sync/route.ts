import { SyncWorkoutDetailsToServerResponse } from "@/api/workoutDetail.api";
import { handleServerError, HttpError } from "@/app/api/_utils/handleError";
import { prisma } from "@/lib/prisma";
import { localWorkoutDetailSchema } from "@/types/models";
import { validateData } from "@/util/validateData";
import { NextRequest, NextResponse } from "next/server";
import pMap from "p-map";
import { z } from "zod";

const requestBodySchema = z.object({
  mappedUnsynced: z.array(
    localWorkoutDetailSchema.extend({
      workoutId: z.string(),
    })
  ),
});

type RequestBody = z.infer<typeof requestBodySchema>;

export const POST = async (req: NextRequest) => {
  try {
    const body = await req.json();
    const parsedBody = validateData<RequestBody>(requestBodySchema, body);

    const unsynced = parsedBody.mappedUnsynced;

    const updatedList: SyncWorkoutDetailsToServerResponse["updated"] = [];
    await pMap(
      unsynced,
      async (detail) => {
        const localId = detail.id;

        if (!localId) throw new HttpError("localId가 없습니다", 422);

        const {
          createdAt,
          updatedAt,
          id,
          isSynced,
          exerciseName,
          serverId,
          workoutId,
          exerciseId,

          ...detailInput
        } = detail;

        let serverDetailId = serverId;
        if (serverDetailId) {
          const updated = await prisma.workoutDetail.update({
            where: { id: serverDetailId },
            data: {
              ...detailInput,
            },
          });
          serverDetailId = updated.id;
        } else {
          const workout = await prisma.workout.findUnique({
            where: { id: workoutId },
          });
          if (!workout) {
            throw new HttpError(
              `workoutId가 일치하는 workout을 찾지 못했습니다`,
              404
            );
          }

          const created = await prisma.workoutDetail.upsert({
            where: {
              workoutId_exerciseOrder_setOrder: {
                workoutId: workout.id,
                exerciseOrder: detail.exerciseOrder,
                setOrder: detail.setOrder,
              },
            },
            update: {
              ...detailInput,
            },
            create: {
              ...detailInput,
              workout: { connect: { id: workoutId } },
              exercise: { connect: { id: exerciseId } },
            },
          });
          serverDetailId = created.id;
        }
        updatedList.push({
          localId,
          serverId: serverDetailId,
          exerciseId,
          workoutId,
        });
      },
      { concurrency: 5 }
    );

    return NextResponse.json({ success: true, updated: updatedList });
  } catch (e) {
    return handleServerError(e);
  }
};
