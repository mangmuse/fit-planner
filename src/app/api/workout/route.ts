import { getWorkouts } from "@/app/api/_utils/getWorkouts";
import { handleServerError } from "@/app/api/_utils/handleError";
import { NextRequest, NextResponse } from "next/server";

export const GET = async (request: NextRequest) => {
  try {
    const searchParams = request.nextUrl.searchParams;
    const userId = searchParams.get("userId");
    if (!userId) {
      return NextResponse.json(
        { success: false, message: "userId가 필요합니다" },
        { status: 400 }
      );
    }
    const workouts = await getWorkouts(userId);

    return NextResponse.json({ success: true, workouts }, { status: 200 });
  } catch (e) {
    return handleServerError(e);
  }
};
