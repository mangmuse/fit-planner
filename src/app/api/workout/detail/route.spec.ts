jest.mock("@/app/api/_utils/getWorkouts");
jest.mock("@/app/api/_utils/handleError");
jest.mock("@/util/validateData");

import { createPrismaWorkoutDetailResponse } from "@/__mocks__/app/api/workout/detail/mockData";
import { getWorkoutIds } from "@/app/api/_utils/getWorkouts";
import { handleServerError, HttpError } from "@/app/api/_utils/handleError";
import { GET, WorkoutDetailWithExercise } from "@/app/api/workout/detail/route";
import { prisma } from "@/lib/prisma";
import { ApiErrorResponse, ApiSuccessResponse } from "@/types/apiResponse";
import { ClientWorkoutDetail } from "@/types/models";
import { validateData } from "@/util/validateData";
import { NextRequest, NextResponse } from "next/server";

describe("GET /api/workout/detail", () => {
  const mockedGetWorkoutIds = getWorkoutIds as jest.Mock;
  const mockedValidateData = validateData as jest.Mock;
  const mockedHandleServerError = handleServerError as jest.Mock;
  const mockedPrismaFindMany = prisma.workoutDetail.findMany as jest.Mock;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const targetUrl = "http://localhost/api/workout/detail";
  const mockUserId = "user-123";
  const mockFindManyResponse: WorkoutDetailWithExercise[] = [
    createPrismaWorkoutDetailResponse({ workoutId: "workout-1" }),
    createPrismaWorkoutDetailResponse({ workoutId: "workout-2" }),
    createPrismaWorkoutDetailResponse({ workoutId: "workout-3" }),
  ];

  it("workoutDetail이 없는경우 200 상태코드와 함께 빈 배열을 반환한다", async () => {
    const req = new NextRequest(`${targetUrl}?userId=${mockUserId}`);

    mockedValidateData.mockReturnValue(mockUserId);
    mockedGetWorkoutIds.mockResolvedValue([]);
    mockedPrismaFindMany.mockResolvedValue([]);

    const res = await GET(req);
    const body = (await res.json()) as ApiSuccessResponse<{
      workoutDetails: ClientWorkoutDetail[];
    }>;

    expect(res.status).toBe(200);
    expect(body.workoutDetails).toEqual([]);
  });

  it("userId가 없거나 올바르지 않은 타입이면 400 상태코드와 함께 에러 메시지를 반환한다", async () => {
    const req = new NextRequest(`${targetUrl}`);

    const res = await GET(req);
    const body = (await res.json()) as ApiErrorResponse;

    expect(res.status).toBe(400);
    expect(body.message).toBe("userId가 없거나 타입이 올바르지 않습니다");

    expect(mockedGetWorkoutIds).not.toHaveBeenCalled();
    expect(prisma.workoutDetail.findMany).not.toHaveBeenCalled();
    expect(mockedHandleServerError).not.toHaveBeenCalled();
  });

  it("DB 조회 중 오류가 발생하면 500 상태코드와 함께 에러 메시지를 반환한다", async () => {
    const req = new NextRequest(`${targetUrl}?userId=${mockUserId}`);
    const mockError = new Error("DB 조회 중 오류");

    mockedValidateData.mockReturnValue(mockUserId);
    mockedGetWorkoutIds.mockResolvedValue([
      "workout-1",
      "workout-2",
      "workout-3",
    ]);

    mockedPrismaFindMany.mockRejectedValue(mockError);
    mockedHandleServerError.mockResolvedValue(
      NextResponse.json(
        {
          success: false,
          message: mockError.message,
        },
        {
          status: 500,
        }
      )
    );

    const res = await GET(req);
    const body = (await res.json()) as ApiErrorResponse;

    expect(res.status).toBe(500);
    expect(body.message).toBe("DB 조회 중 오류");

    expect(mockedGetWorkoutIds).toHaveBeenCalledWith(mockUserId);
    expect(mockedHandleServerError).toHaveBeenCalled();
  });

  it("userId가 올바른 타입이면 200 상태코드와 함께 운동 상세 목록을 반환한다", async () => {
    const req = new NextRequest(`${targetUrl}?userId=${mockUserId}`);

    mockedValidateData.mockReturnValue(mockUserId);
    mockedGetWorkoutIds.mockResolvedValue([
      "workout-1",
      "workout-2",
      "workout-3",
    ]);

    mockedPrismaFindMany.mockResolvedValue(mockFindManyResponse);

    const res = await GET(req);
    const body = (await res.json()) as ApiSuccessResponse<{
      workoutDetails: ClientWorkoutDetail[];
    }>;

    expect(res.status).toBe(200);
    expect(body.success).toBe(true);
    expect(body.workoutDetails).toMatchSnapshot();
    expect(mockedGetWorkoutIds).toHaveBeenCalledWith(mockUserId);
    expect(mockedHandleServerError).not.toHaveBeenCalled();
  });
});
