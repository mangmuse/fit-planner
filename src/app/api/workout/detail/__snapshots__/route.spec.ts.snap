// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api/workout/detail userId가 올바른 타입이면 200 상태코드와 함께 운동 상세 목록을 반환한다 1`] = `
[
  {
    "createdAt": "2024-01-01T10:00:00.000Z",
    "exerciseId": 1,
    "exerciseName": "벤치프레스",
    "exerciseOrder": 1,
    "id": "workout-detail-1",
    "isDone": false,
    "reps": 10,
    "rpe": null,
    "setOrder": 1,
    "setType": "NORMAL",
    "updatedAt": "2024-01-01T10:00:00.000Z",
    "weight": 60,
    "weightUnit": "kg",
    "workoutId": "workout-1",
  },
  {
    "createdAt": "2024-01-01T10:00:00.000Z",
    "exerciseId": 1,
    "exerciseName": "벤치프레스",
    "exerciseOrder": 1,
    "id": "workout-detail-1",
    "isDone": false,
    "reps": 10,
    "rpe": null,
    "setOrder": 1,
    "setType": "NORMAL",
    "updatedAt": "2024-01-01T10:00:00.000Z",
    "weight": 60,
    "weightUnit": "kg",
    "workoutId": "workout-2",
  },
  {
    "createdAt": "2024-01-01T10:00:00.000Z",
    "exerciseId": 1,
    "exerciseName": "벤치프레스",
    "exerciseOrder": 1,
    "id": "workout-detail-1",
    "isDone": false,
    "reps": 10,
    "rpe": null,
    "setOrder": 1,
    "setType": "NORMAL",
    "updatedAt": "2024-01-01T10:00:00.000Z",
    "weight": 60,
    "weightUnit": "kg",
    "workoutId": "workout-3",
  },
]
`;
